src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +119,242 @@ def snapshot(self) -> Iterable[JobInfo]:
+        super().__init__(sbe, "hive_metastore", schema, "submit_runs", SubmitRunInfo)
+        self._ws = ws
+        self._num_days_history = num_days_history
+
+    def snapshot(self) -> Iterable[SubmitRunInfo]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    @staticmethod
+    def dt_to_ms(dt: datetime):
+        epoch = datetime.utcfromtimestamp(0)
+        delta = dt - epoch
+        return int(delta.total_seconds() * 1000)
+
+    @staticmethod
+    def get_current_dttm() -> datetime:
+        return datetime.now()
+
+    def _crawl(self) -> Iterable[SubmitRunInfo]:
+        end = self.get_current_dttm()
+        start = end - timedelta(days=self._num_days_history)
+        submit_runs = self._ws.jobs.list_runs(
+            expand_tasks=True,
+            completed_only=True,
+            run_type=ListRunsRunType.SUBMIT_RUN,
+            start_time_from=self.dt_to_ms(start),
+            start_time_to=self.dt_to_ms(end),
+        )
+        return self.assess_job_runs(submit_runs)
+
+    def _try_fetch(self) -> Iterable[SubmitRunInfo]:
+        for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
+            yield SubmitRunInfo(*row)
+
+    @staticmethod
+    def retrieve_hash_values_from_sql_task(task: SqlTask) -> list[str]:
+        hash_values = []
+        if task.file is not None:
+            hash_values.append(task.file.path)
+        if task.alert is not None and task.alert.alert_id is not None:
+            hash_values.append(task.alert.alert_id)
+        if task.dashboard is not None and task.dashboard.dashboard_id is not None:
+            hash_values.append(task.dashboard.dashboard_id)
+        if task.query is not None and task.query.query_id is not None:
+            hash_values.append(task.query.query_id)
+        return [str(value) for value in hash_values if value is not None]
+
+    @staticmethod
+    def retrieve_hash_values_from_git_source(source: GitSource) -> list[str]:
+        hash_values = []
+        if source.git_url is not None:
+            hash_values.append(source.git_url)
+        return [str(value) for value in hash_values if value is not None]
+
+    @staticmethod
+    def retrieve_hash_values_from_dbt_task(dbt_task: DbtTask) -> list[str]:
+        hash_values = []
+        dbt_task.commands.sort()
+        if dbt_task.schema is not None:
+            hash_values.append(dbt_task.schema)
+        if dbt_task.catalog is not None:
+            hash_values.append(dbt_task.catalog)
+        if dbt_task.warehouse_id is not None:
+            hash_values.append(dbt_task.warehouse_id)
+        if dbt_task.project_directory is not None:
+            hash_values.append(dbt_task.project_directory)
+        hash_values.append(",".join(dbt_task.commands))
+        return [str(value) for value in hash_values if value is not None]
+
+    @staticmethod
+    def retrieve_hash_values_from_jar_task(spark_jar_task: SparkJarTask) -> list[str]:
```

Comment:
make private. you seem to have a wrong understanding of object oriented design principles, especially related to private/public methods. 

all methods have to be private by default. only those methods called by other classes can be public. we have to be critical about the number of public members, as it may cause maintainability issues. you don't just make a method public for the sake of unit testing.